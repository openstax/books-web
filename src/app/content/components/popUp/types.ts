import { LinkedArchiveTree, LinkedArchiveTreeNode, LinkedArchiveTreeSection } from '../../types';

export type LocationFilters = Map<
  string,
  { section: LinkedArchiveTreeNode, children?: LinkedArchiveTreeSection[] }
>;

export type LocationFiltersWithChildren = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;

export interface FiltersChange<T> {
  remove: T[];
  new: T[];
}
