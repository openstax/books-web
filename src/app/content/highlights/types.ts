import { Highlight, HighlightsSummary } from '@openstax/highlighter/dist/api';

export type HighlightData = Highlight;

export type CountsPerSource = Exclude<HighlightsSummary['countsPerSource'], undefined>;

export interface State {
  myHighlightsOpen: boolean;
  enabled: boolean;
  focused?: string;
  highlights: null | HighlightData[];
  summary: {
    filters: {
      colors: string[];
      chapters: string[];
    },
    pagination: {
      // even though we're manually splitting our requests out into smaller batches of sources,
      // if a source has more than [numPage] highlights the api will still paginate. in order to
      // access the subsequent pages, we need to store the original sources here, and continue to
      // use them until we run out of pages, then grab the next batch of sources. on the LAST
      // page the response size may be smaller than expected, in this case the next batch of
      // sources should be immediately queried to fill the remaining [numPage]. use
      // paginationUtils.getNextPageSources to get the next batch of sources
      sources: string[];
      page: number;
    } | null,
    loading: boolean;
    // totalCounts should reflect the UNFILTERED state of user's data, used in representing
    // available filter options. this should be updated on create/delete and after auth change
    totalCounts: {[key: string]: number};
    // filteredTotalCounts should reflect the FILTERED state of user's data, used in calculating
    // pagination requests. this should be updated after filter change before requesting
    // filtered highlights (because you need this to query them)
    filteredTotalCounts: {[key: string]: number};
    highlights: {
      [chapterId: string]: {[pageId: string]: HighlightData[]}
    };
  };
}
