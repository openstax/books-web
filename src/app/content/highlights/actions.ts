import { createStandardAction } from 'typesafe-actions';

export const enableHighlighting = createStandardAction('FeatureFlag/enableHighlighting')<void>();
