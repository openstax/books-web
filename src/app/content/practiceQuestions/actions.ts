import { createStandardAction } from 'typesafe-actions';
import { PracticeQuestionsSummary } from './types';

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();
