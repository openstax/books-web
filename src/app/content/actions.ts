import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as searchActions from './search/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();

export const closeCallToActionPopup = createStandardAction('Content/closeCallToActionPopup')();
export const openCallToActionPopup = createStandardAction('Content/openCallToActionPopup')();
