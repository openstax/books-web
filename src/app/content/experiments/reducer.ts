import { Reducer } from 'redux';
import { getType } from 'typesafe-actions';
import { receiveFeatureFlags } from '../../actions';
import { AnyAction } from '../../types';
import { experimentIds } from './constants';

export const initialState = {};

const reducer: Reducer<{}, AnyAction> = (state = initialState, action): any => {
    switch (action.type) {
        case getType(receiveFeatureFlags):
            const [id, variant] = action.payload;
            if (experimentIds[id]) {
                const experiment = experimentIds[id];
                const idx = parseInt(variant, 10);
                return {...state, [experiment]: idx};
            }
            return state;
        default:
            return state;
    }
};

export default reducer;
