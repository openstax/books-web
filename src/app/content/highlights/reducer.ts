import { Highlight, HighlightColorEnum, HighlightSourceTypeEnum } from '@openstax/highlighter/dist/api';
import omit from 'lodash/fp/omit';
import { Reducer } from 'redux';
import { getType } from 'typesafe-actions';
import { receiveFeatureFlags } from '../../actions';
import { locationChange } from '../../navigation/actions';
import { AnyAction } from '../../types';
import { merge } from '../../utils';
import * as actions from './actions';
import { highlightingFeatureFlag, highlightStyles } from './constants';
import { State } from './types';
import {
  addToTotalCounts,
  getHighlightByIdFromSummaryHighlights,
  getHighlightColorFiltersWithContent,
  getHighlightLocationFiltersWithContent,
  removeFromTotalCounts,
  removeSummaryHighlight,
  updateInTotalCounts,
  updateSummaryHighlightsDependOnFilters
} from './utils';

const defaultColors = highlightStyles.map(({label}) => label);
export const initialState: State = {
  enabled: false,
  hasUnsavedHighlight: false,
  highlights: null,
  myHighlightsOpen: false,
  summary: {
    filters: {colors: defaultColors, locationIds: []},
    highlights: null,
    loading: false,
    pagination: null,
    totalCountsPerPage: null,
  },
};

const reducer: Reducer<State, AnyAction> = (state = initialState, action) => {
  switch (action.type) {
    case getType(receiveFeatureFlags): {
      return {...state, enabled: action.payload.includes(highlightingFeatureFlag)};
    }
    case getType(locationChange): {
      return {...initialState, enabled: state.enabled, myHighlightsOpen: false,
        summary: {...state.summary, loading: true},
      };
    }
    case getType(actions.createHighlight): {
      const highlight = {
        ...action.payload,
        color: action.payload.color as string as HighlightColorEnum,
        sourceType: action.payload.sourceType as string as HighlightSourceTypeEnum,
      };

      const newSummaryHighlights = state.summary.highlights
        ? updateSummaryHighlightsDependOnFilters(
          state.summary.highlights,
          state.summary.filters,
          {...action.meta, highlight })
        : state.summary.highlights;

      const totalCountsPerPage = state.summary.totalCountsPerPage
        ? addToTotalCounts(state.summary.totalCountsPerPage, highlight)
        : state.summary.totalCountsPerPage;

      return {
        ...state,
        hasUnsavedHighlight: false,
        highlights: [...state.highlights || [], highlight],
        summary: {
          ...state.summary,
          highlights: newSummaryHighlights || state.summary.highlights,
          totalCountsPerPage,
        },
      };
    }
    case getType(actions.openMyHighlights):
      return {...state, myHighlightsOpen: true};
    case getType(actions.closeMyHighlights):
      return {...state, myHighlightsOpen: false};
    case getType(actions.updateHighlight): {
      if (!state.highlights) { return state; }

      const oldHighlight = state.highlights.find((highlight) => highlight.id === action.payload.id)
        || getHighlightByIdFromSummaryHighlights(state.summary.highlights || {}, action.payload.id);
      if (!oldHighlight) { return state; }

      const hasUnsavedHighlight =
        oldHighlight.annotation === action.payload.highlight.annotation
        && state.hasUnsavedHighlight;

      const newHighlight = {
        ...oldHighlight,
        ...action.payload.highlight,
      } as Highlight;

      const newHighlights = state.highlights.map((highlight) => {
        if (highlight.id === oldHighlight.id) { return newHighlight; }
        return highlight;
      });

      const newSummaryHighlights = state.summary.highlights
        ? updateSummaryHighlightsDependOnFilters(
          state.summary.highlights,
          state.summary.filters,
          {...action.meta, highlight: newHighlight})
        : state.summary.highlights
      ;

      const totalCountsPerPage = state.summary.totalCountsPerPage
        ? updateInTotalCounts(state.summary.totalCountsPerPage, oldHighlight, newHighlight)
        : state.summary.totalCountsPerPage
      ;

      return {
        ...state,
        hasUnsavedHighlight: Boolean(hasUnsavedHighlight),
        highlights: newHighlights,
        summary: {
          ...state.summary,
          highlights: newSummaryHighlights,
          totalCountsPerPage,
        },
      };
    }
    case getType(actions.deleteHighlight): {
      if (!state.highlights) {
        return state;
      }

      const [newSummaryHighlights, removedHighlight] = state.summary.highlights
        ? removeSummaryHighlight(state.summary.highlights, {
          ...action.meta,
          id: action.payload,
        })
        : [state.summary.highlights, null]
      ;

      const totalCountsPerPage = state.summary.totalCountsPerPage && removedHighlight
        ? removeFromTotalCounts(state.summary.totalCountsPerPage, removedHighlight)
        : state.summary.totalCountsPerPage
      ;

      return {
        ...state,
        focused: state.focused === action.payload ? undefined : state.focused,
        hasUnsavedHighlight: false,
        highlights: state.highlights.filter(({id}) => id !== action.payload),
        summary: {
          ...state.summary,
          highlights: newSummaryHighlights,
          totalCountsPerPage,
        },
      };
    }
    case getType(actions.receiveHighlights): {
      return {...state, highlights: [...state.highlights || [], ...action.payload],
        summary: {...state.summary, loading: false},
      };
    }
    case getType(actions.focusHighlight): {
      return {...state, focused: action.payload, hasUnsavedHighlight: false };
    }
    case getType(actions.clearFocusedHighlight): {
      return {
        ...omit('focused', state),
        hasUnsavedHighlight: false,
      };
    }
    case getType(actions.setAnnotationChangesPending): {
      return {
        ...state,
        hasUnsavedHighlight: action.payload,
      };
    }
    case getType(actions.initializeMyHighlightsSummary):
    case getType(actions.loadMoreSummaryHighlights): {
      return {
        ...state,
        summary: {
          ...state.summary,
          loading: true,
        },
      };
    }
    case getType(actions.setSummaryFilters): {
      return {
        ...state,
        summary: {
          ...state.summary,
          filters: {
            ...state.summary.filters,
            ...action.payload,
          },
          highlights: {},
          loading: true,
          pagination: null,
        },
      };
    }
    case getType(actions.receiveSummaryHighlights): {
      return {
        ...state,
        summary: {
          ...state.summary,
          highlights: merge(state.summary.highlights || {}, action.payload),
          loading: false,
          pagination: action.meta,
        },
      };
    }
    case getType(actions.receiveHighlightsTotalCounts): {
      const locationIds = Array.from(getHighlightLocationFiltersWithContent(action.meta, action.payload));
      const colors = Array.from(getHighlightColorFiltersWithContent(action.payload));

      return {
        ...state,
        summary: {
          ...state.summary,
          filters: {
            colors,
            locationIds,
          },
          totalCountsPerPage: action.payload,
        },
      };
    }
    default:
      return state;
  }
};

export default reducer;
