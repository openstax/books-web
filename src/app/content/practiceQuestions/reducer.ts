import { Reducer } from 'redux';
import { getType } from 'typesafe-actions';
import { receiveFeatureFlags } from '../../actions';
import { AnyAction } from '../../types';
import { practiceQuestionsFeatureFlag } from '../constants';
import { closePracticeQuestions, openPracticeQuestions } from './actions';
import { State } from './types';

export const initialState: State = {
  isEnabled: false,
  open: false,
};

const reducer: Reducer<State, AnyAction> = (state = initialState, action) => {
  switch (action.type) {
    case getType(receiveFeatureFlags):
      return {...state, isEnabled: action.payload.includes(practiceQuestionsFeatureFlag)};
    case getType(openPracticeQuestions):
      return {...state, open: true};
    case getType(closePracticeQuestions):
      return {...state, open: false};
    default:
      return state;
  }
};

export default reducer;
