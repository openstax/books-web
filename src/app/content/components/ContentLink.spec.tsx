import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { push } from '../../navigation/actions';
import { AppState } from '../../types';
import { initialState } from '../reducer';
import { content } from '../routes';
import ConnectedContentLink from './ContentLink';

const BOOK_SLUG = 'bookslug';
const PAGE_SLUG = 'page-title';

const book = {
  id: 'booklongid',
  shortId: 'book',
  slug: BOOK_SLUG,
  title: 'book title',
  tree: {
    contents: [
      {
        id: 'pagelongid@0',
        shortId: 'page@0',
        title: 'page title',
      },
    ],
    id: 'booklongid@0',
    shortId: 'book@0',
    title: 'book title',
  },
  version: '0',
};
const page = {
  id: 'pagelongid',
  shortId: 'page',
  title: 'page title',
  version: '0',
};

describe('ContentLink', () => {
  let consoleError: jest.SpyInstance;

  beforeEach(() => {
    consoleError = jest.spyOn(console, 'error');
  });

  afterEach(() => {
    consoleError.mockRestore();
  });

  it('dispatches navigation action on click', () => {
    const pathname = '/doesnotmatter';
    const state = {
      content: {
        ...initialState,
        book, page,
      },
      navigation: { pathname },
    } as any as AppState;
    const store = createStore((s: AppState | undefined) => s || state, state);
    const dispatchSpy = jest.spyOn(store, 'dispatch');
    const component = renderer.create(<Provider store={store}>
      <ConnectedContentLink book={book} page={page} />
    </Provider>);
  
    const event = {
      preventDefault: jest.fn(),
    };

    component.root.findByType('a').props.onClick(event);

    expect(dispatchSpy).toHaveBeenCalledWith(push({
      params: {book: BOOK_SLUG, page: PAGE_SLUG},
      route: content,
      state: {
        bookUid: 'booklongid',
        bookVersion: '0',
        pageUid: 'pagelongid',
      },
    }));
    expect(event.preventDefault).toHaveBeenCalled();
  });

});
