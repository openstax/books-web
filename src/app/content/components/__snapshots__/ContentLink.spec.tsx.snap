// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentLink creates relative paths when at the root (unused) 1`] = `
<a
  href="books/bookslug/pages/page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when deeply under the same Page (unused) 1`] = `
<a
  href=".."
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when in a different book 1`] = `
<a
  href="../../bookslug/pages/page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when in the same book 1`] = `
<a
  href="page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when not in a book and not at the root (unused) 1`] = `
<a
  href="../books/bookslug/pages/page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when not in a book and not at the root (unused) 2`] = `
<a
  href="../books/bookslug/pages/page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when the current path ends in a / (unused) 1`] = `
<a
  href="../page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when the current path ends in a / but is not the current page (unused) 1`] = `
<a
  href="../page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when the same page 1`] = `
<a
  href="page-title"
  onClick={[Function]}
/>
`;

exports[`ContentLink creates relative paths when under the same Page (unused) 1`] = `
<a
  href="../page-title"
  onClick={[Function]}
/>
`;
